Linux/sgx/test_app/enclave/TestEnclave.cpp

```diff
@@ -90,184 +90,193 @@ struct evp_pkey_st {
 
 void rsa_key_gen()
 {
-	BIGNUM *bn = BN_new();
-	if (bn == NULL) {
-		printf("BN_new failure: %ld\n", ERR_get_error());
-	    return;
-	}
-	int ret = BN_set_word(bn, RSA_F4);
-    if (!ret) {
-       	printf("BN_set_word failure\n");
-	    return;
-	}
-	
-	RSA *keypair = RSA_new();
-	if (keypair == NULL) {
-		printf("RSA_new failure: %ld\n", ERR_get_error());
-	    return;
-	}
-	ret = RSA_generate_key_ex(keypair, 4096, bn, NULL);
-	if (!ret) {
-        printf("RSA_generate_key_ex failure: %ld\n", ERR_get_error());
-	    return;
-	}
-
-	EVP_PKEY *evp_pkey = EVP_PKEY_new();
-	if (evp_pkey == NULL) {
-		printf("EVP_PKEY_new failure: %ld\n", ERR_get_error());
-		return;
-	}
-	EVP_PKEY_assign_RSA(evp_pkey, keypair);
-
-	// public key - string
-	int len = i2d_PublicKey(evp_pkey, NULL);
-	unsigned char *buf = (unsigned char *) malloc (len + 1);
-	unsigned char *tbuf = buf;
-	i2d_PublicKey(evp_pkey, &tbuf);
-
-	// print public key
-	printf ("{\"public\":\"");
-	int i;
-	for (i = 0; i < len; i++) {
-	    printf("%02x", (unsigned char) buf[i]);
-	}
-	printf("\"}\n");
-
-	free(buf);
-
-	// private key - string
-	len = i2d_PrivateKey(evp_pkey, NULL);
-	buf = (unsigned char *) malloc (len + 1);
-	tbuf = buf;
-	i2d_PrivateKey(evp_pkey, &tbuf);
-
-	// print private key
-	printf ("{\"private\":\"");
-	for (i = 0; i < len; i++) {
-	    printf("%02x", (unsigned char) buf[i]);
-	}
-	printf("\"}\n");
-
-	free(buf);
-
-	BN_free(bn);
-
-	EVP_PKEY_free(evp_pkey);
-
-	if (evp_pkey->pkey.ptr != NULL) {
-	  RSA_free(keypair);
-	}
+    EVP_PKEY_CTX* ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_RSA, NULL);
+    if (!ctx)
+    {
+        printf("EVP_PKEY_CTX_new_id: %ld\n", ERR_get_error());
+        EVP_PKEY_CTX_free(ctx);
```

Comment:
Seems no need to free if allocation fails.

Reply:
Fixed